src/main/java/com/entropyteam/entropay/employees/services/ReportService.java

```diff
@@ -255,9 +254,56 @@ private List<PtoReportDto> getPtoReportDtos(List<Pto> ptoList) {
                             .map(assignment -> assignment.getProject().getClient().getName()).orElse("No client");
                     int totalPtoDays = totalPtoDaysMap.getOrDefault(employee.getId(), 0);
 
-                    return new PtoReportDto(employee.getId(), employee.getInternalId(), employee.getFirstName(), employee.getLastName(),
-                            clientName, totalPtoDays, 0);
+                    return new PtoReportEmployeeDto(employee.getId(), employee.getInternalId(), employee.getFirstName(), employee.getLastName(),
+                            clientName, totalPtoDays, year);
                 })
                 .collect(Collectors.toList());
     }
+
+    public Page<PtoReportClientDto> getPtosReportByClients(ReactAdminParams params) {
+        List<PtoReportClientDto> ptoReportDtoList;
+        Filter filter = mapper.buildReportFilter(params, PtoReportClientDto.class);
+        if (filter.getGetByFieldsFilter().containsKey(YEAR)) {
+            Integer year = getYearFromFilter(filter);
+            ptoReportDtoList = getPtoReportClientDtos(ptoRepository.findAllByDeletedIsFalseAndStatusIsForYear(Status.APPROVED.name(), year), year);
+        }
+        else {
+            Integer year = LocalDate.now().getYear();
+            ptoReportDtoList = getPtoReportClientDtos(ptoRepository.findAllByDeletedIsFalseAndStatusIsForYear(Status.APPROVED.name(), year), year);
+        }
+        return new PageImpl<>(ptoReportDtoList, Pageable.unpaged(), ptoReportDtoList.size());
+    }
+
+    private List<PtoReportClientDto> getPtoReportClientDtos(List<Pto> ptoList, Integer year) {
+        List<Client> clients = clientRepository.findAllClientsWithAProject();
+        List<Employee> employeeList = employeeRepository.findAllByIdInAndDeletedIsFalse(ptoList.stream().map(pto -> pto.getEmployee().getId()).collect(Collectors.toList()));
+
+        Map<UUID, List<Assignment>> assignmentMap = assignmentRepository.findAllAssignmentsByClientIdIn(clients.stream().map(BaseEntity::getId).collect(Collectors.toList()))
+                .stream()
+                .collect(Collectors.groupingBy(assignment -> assignment.getProject().getClient().getId()));
+
+        return clients.stream().map(client -> {
+            List<Assignment> clientAssignments = assignmentMap.getOrDefault(client.getId(), Collections.emptyList());
+            List<Employee> clientEmployees = employeeList.stream()
+                    .map(Employee::getId)
+                    .flatMap(employeeId -> clientAssignments.stream()
+                            .map(Assignment::getEmployee)
+                            .filter(employee -> employee.getId().equals(employeeId)))
+                    .distinct()
+                    .toList();
```

Comment:
I think you can avoid these loops, we already have the employee id in the assignment